x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,34 +105,126 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
         }
     }
 
-    private void readInternalBytes(final long part, final long pos, byte[] b, int offset, int length) throws IOException {
+    private void readInternalBytes(final int part, long pos, final byte[] b, int offset, int length) throws IOException {
+        final long currentSequentialReadSize = sequentialReadSize;
+        if (currentSequentialReadSize != NO_SEQUENTIAL_READ_OPTIMIZATION) {
+            final StreamForSequentialReads streamForSequentialReads = streamForSequentialReadsRef.get();
+            if (streamForSequentialReads == null) {
+                // start a new sequential read
+                if (tryReadAndKeepStreamOpen(part, pos, b, offset, length, currentSequentialReadSize)) {
+                    return;
+                }
+            } else if (streamForSequentialReads.part == part && streamForSequentialReads.pos == pos) {
+                // continuing a sequential read that we started previously
+                assert streamForSequentialReads.isFullyRead() == false;
+                int read = streamForSequentialReads.inputStream.read(b, offset, length);
+                assert read <= length : read + " vs " + length;
+                streamForSequentialReads.pos += read;
```

Comment:
should we put this logic into StreamForSequentialReads? Perhaps that class could enforce that only sequential reads are possible from the stream (and offer a method to say isSequentialReadPossible), with the logic in this class here just trying to call these methods.

Reply:
Moved this around in 1d692930391c32e0ca4f2bf0fd6d0e6fbbce7ec6 and 184c9b13764642baa3a6daf9ccabce54e4ef24d5.